### PR TITLE
Upgrade labels check script

### PR DESCRIPTION
**Pull request by QuantumEnigmaa**
Towards: https://github.com/giantswarm/giantswarm/issues/24956

This PR upgrades the go script checking for missing labels in alerting rules by using the alertmanager config file.

### Checklist

- [ ] Update CHANGELOG.md
- [ ] Add [Unit tests](https://github.com/giantswarm/prometheus-rules/#testing)
- [ ] Follow [Alert structure](https://github.com/giantswarm/prometheus-rules/#how-alerts-are-structured)
- [ ] Consider [creating a dashboard](https://docs.giantswarm.io/getting-started/observability/grafana/custom-dashboards/) ([guidelines](https://intranet.giantswarm.io/docs/product/ux/guidelines/dashboards/)) (if it does not exist already) to help oncallers monitor the status of the issue.
- [ ] Request review from oncall area, as well as team (e.g: `oncall-kaas-cloud` GitHub group).

### CODE REVIEW

**Comment by QuantumEnigmaa** [comment]
As one can see in the test / gitleaks github action, the script is working fine :)

**Review by QuentinBisson — APPROVED** [review]
LGTM

File: test/hack/bin/get-inhibition.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+GIT_WORKDIR="$(git rev-parse --show-toplevel)"
+REPO_VERSION="$(curl -s https://api.github.com/repos/giantswarm/prometheus-meta-operator/releases/latest | jq -r .name)"
+
+curl https://raw.githubusercontent.com/giantswarm/prometheus-meta-operator/"$REPO_VERSION"/files/templates/alertmanager/alertmanager.yaml > "$GIT_WORKDIR"/test/hack/checkLabels/alertmanager.yaml
```
Comment (hervenicol) [inline test/hack/bin/get-inhibition.sh:6]: I don't think the progress bar is very useful here; I'd rather have it less verbose.
```suggestion
curl -s https://raw.githubusercontent.com/giantswarm/prometheus-meta-operator/"$REPO_VERSION"/files/templates/alertmanager/alertmanager.yaml > "$GIT_WORKDIR"/test/hack/checkLabels/alertmanager.yaml
```

File: test/hack/checkLabels/main.go [hunk 2]
```diff
@@ -103,32 +143,57 @@ func getMissingLabels() ([]string, error) {
...
 func main() {
-	missingLabels, err := getMissingLabels()
+	missingTargetLabels, missingCancelLabels, err := getMissingLabels()
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	fmt.Println("MISSING LABELS:")
-	for _, label := range missingLabels {
+	fmt.Println("MISSING TARGET LABELS:")
```
Comment (hervenicol) [inline test/hack/checkLabels/main.go:180]: This message was not clear to me when I ran it locally.
What about:
```suggestion
	fmt.Println("MISSING TARGET LABELS IN ALERTMANAGER INHIBITIONS DEFINITIONS:")
```

**Review by hervenicol — COMMENTED** [review]
Here is what I see when I run `make test-inhibitions`:

```
./test/hack/bin/fetch-tools.sh
# prepare the helm chart
test/hack/bin/architect helm template --dir helm/prometheus-rules --dry-run
2023/01/10 12:07:50 templating helm chart
dir: helm/prometheus-rules
sha: 305f7cf103e112a89ba2fbc783e4800e4d10ad05
tag: 
app-version: 2.72.0-305f7cf103e112a89ba2fbc783e4800e4d10ad05
version: 2.72.0-305f7cf103e112a89ba2fbc783e4800e4d10ad05
2023/01/10 12:07:50 templated helm chart
# test whether inhibition labels are well defined
./test/hack/bin/get-inhibition.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14139  100 14139    0     0   180k      0 --:--:-- --:--:-- --:--:--  181k
./test/hack/bin/template-chart.sh
./test/hack/bin/template-chart.sh: line 3: local: can only be used in a function
./test/hack/bin/template-chart.sh: line 6: local: can only be used in a function
Templating chart for provider: openstack
wrote /home/herve/github/giantswarm/prometheus-rules/test/hack/output/openstack/prometheus-rules/templates/alerting-rules/alertmanager.rules.yml
wrote /home/herve/github/giantswarm/prometheus-rules/test/hack/output/openstack/prometheus-rules/templates/alerting-rules/apiserver.management-cluster.rules.yml
wrote /home/herve/github/giantswarm/prometheus-rules/test/hack/output/openstack/prometheus-rules/templates/alerting-rules/apiserver.workload-cluster.rules.yml
[...]
[... a ton of these lines]
[...]
wrote /home/herve/github/giantswarm/prometheus-rules/test/hack/output/capz/prometheus-rules/templates/recording-rules/service-level.rules.yml

cd test/hack/checkLabels; go run main.go
MISSING TARGET LABELS:
cancel_if_stack_failed

MISSING SOURCE LABELS:
cluster_id
kubelet_not_ready
scrape_timeout
apiserver_down
```

My remarks:
- It's very verbose, I'm not sure all of this information is useful
- template-chart.sh has some errors (though not related to this PR)
- I don't think `cluster_id` should be part of the list of missing source labels, as it's part of the actual metrics.

**Comment by marieroque** [comment]
It could be great to refer the rules where labels are missing in the script output

File: test/hack/checkLabels/main.go (hunk 2 above)
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go]: Added

File: test/hack/bin/get-inhibition.sh (hunk 1 above)
Comment (QuantumEnigmaa) [reply test/hack/bin/get-inhibition.sh]: Added

**Comment by QuantumEnigmaa** [comment]
I'm not sure it is really necessary @marieroque since one just has to search across all files to find the related labels

File: test/hack/checkLabels/main.go [hunk 3]
```diff
@@ -103,32 +143,57 @@ func getMissingLabels() ([]string, error) {
...
 
 	for _, cancelLabel := range cancelLabels {
-		var originLabel = strings.Split(cancelLabel, "cancel_if_")
-		if len(originLabel) < 2 {
-			return nil, fmt.Errorf("No origin label found in %q", cancelLabel)
+		var i = 0
+
+		for _, targetLabel := range targetLabels {
+			if cancelLabel == targetLabel {
+				i++
```
Comment (marieroque) [inline test/hack/checkLabels/main.go:153]: Once i > 0, you can break out from the for loop

**Comment by QuantumEnigmaa** [comment]
@hervenicol Now cluster_id is ignored but this is based on the hypothesis that all source labels will have boolean values.

File: test/hack/checkLabels/main.go (hunk 3 above)
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:157]: True, I added it

File: helm/prometheus-rules/values.yaml [hunk 4]
```diff
@@ -2,8 +2,8 @@ name: prometheus-rules
 namespace: monitoring
 serviceType: managed
 project:
-  branch: '[[ .Branch ]]'
-  commit: '[[ .SHA ]]'
+  branch: 'labels-matching'
+  commit: 'b8922e8fb6802e25942b60693f0e224fad869e12'
```
Comment (hervenicol) [inline helm/prometheus-rules/values.yaml:6]: Did you commit those on purpose?

Comment (QuantumEnigmaa) [reply helm/prometheus-rules/values.yaml]: Absolutely not ! I don't even know when those changes appeared

Comment (QuantumEnigmaa) [reply helm/prometheus-rules/values.yaml]: Edit: I forgot to `make clean` after running local tests

**Review by hervenicol — APPROVED** [review]
LGTM

**Pull request merged**